ouroboros-consensus-test/src/Test/Util/MockDiffSeq.hs

```diff
@@ -0,0 +1,196 @@
+{-# LANGUAGE DeriveAnyClass             #-}
+{-# LANGUAGE DeriveGeneric              #-}
+{-# LANGUAGE DerivingVia                #-}
+{-# LANGUAGE FlexibleInstances          #-}
+{-# LANGUAGE GeneralisedNewtypeDeriving #-}
+{-# LANGUAGE LambdaCase                 #-}
+{-# LANGUAGE MultiParamTypeClasses      #-}
+{-# LANGUAGE RankNTypes                 #-}
+{-# LANGUAGE ScopedTypeVariables        #-}
+{-# LANGUAGE StandaloneDeriving         #-}
+{-# LANGUAGE StandaloneKindSignatures   #-}
+{-# LANGUAGE TupleSections              #-}
+
+module Test.Util.MockDiffSeq (
+    -- * Types
+    MockDiff (..)
+  , MockDiffEntry (..)
+  , MockDiffSeq (..)
+  , MockKeys (..)
+  , MockSlotNo (..)
+  , MockValues (..)
+    -- * Diff sequence operations
+  , mFlush
+  , mForwardValuesAndKeys
+  , mPush
+  , mRollback
+  , mTotalDiff
+  ) where
+
+import           Prelude hiding (splitAt)
+
+import           Control.DeepSeq
+import           Data.Bifunctor
+import           Data.Kind
+import qualified Data.Map.Merge.Strict as MapMerge
+import           Data.Map.Strict (Map)
+import qualified Data.Map.Strict as Map
+import           Data.Monoid (Sum (..))
+import           Data.Semigroup (Max (..))
+import           Data.Sequence (Seq)
+import qualified Data.Sequence as Seq
+import           Data.Set (Set)
+import qualified Data.Set as Set
+import           GHC.Generics (Generic)
+
+import           Data.FingerTree.Strict
+
+import           Test.Util.Orphans.NFData ()
+
+
+{------------------------------------------------------------------------------
+  Types
+------------------------------------------------------------------------------}
+
+type MockDiffSeq :: Type -> Type -> Type
+newtype MockDiffSeq k v = MockDiffSeq {
+    fromMockDiffSeq ::
+      StrictFingerTree
+        (MockLength, MockDiff k v)
+        (MockSlotNo, MockDiff k v)
+  }
+  deriving stock (Show, Eq, Generic)
+  deriving anyclass NFData
+
+newtype MockLength = MockLength { unMockLength :: Int }
+  deriving stock (Show, Eq, Ord, Generic)
+  deriving newtype Num
+  deriving anyclass NFData
+
+newtype MockDiff k v = MockDiff (Map k (Seq (MockDiffEntry v)))
```

Comment:
Why do we need a `Seq` here? Isn't it sufficient to keep track of the last insertion or deletion of a key?

Reply:
Not needed here, but it might be needed in the final prototype.